### PR TITLE
feat(core): pre-fetch provider tools and make tools a native feature

### PR DESCRIPTION
**Pull request by edvald**
**What this PR does / why we need it**:

This adds the notion of tools to the plugin framework, and a command to
allow using and pre-fetching these tools.

The primary motivation was to allow pre-fetching the tools as part of
our container builds, as well as to make it easy for users to do the
same in their custom CI setups with Garden.

The `garden tools` command is also handy since it allows users to easily
use the exact same tools as Garden's providers fetch and use.

To pre-fetch the tools, we add a `garden util fetch-tools` command,
which fetches all tools for the configured providers in the current
project, or for all registered providers if the `--all` flag is set.

Notes:
- We now always fetch the docker binary on use, instead of checking for
  its existence locally. This avoids some code complexity and makes sure
  we always use the expected version. The downside should be mitigated
  by the pre-fetching in built containers, and the fetch-tools command.
- We really need to reduce the number of tools and versions we bundle
  before we release 0.12. The container image sizes are too large atm.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

Any feedback on the design/usability here would be good, and maybe give it a spin, see if it all works as expected.

### CODE REVIEW

File: garden-service/src/util/ext-tools.ts [hunk 1]
```diff
@@ -7,93 +7,193 @@
...
+    this.versionDirname = hashString(this.buildSpec.url, 16)
     this.versionPath = join(this.toolPath, this.versionDirname)
 
-    this.targetSubpath = this.spec.extract ? this.spec.extract.targetPath : [basename(this.spec.url)]
+    this.targetSubpath = this.buildSpec.extract ? this.buildSpec.extract.targetPath : basename(this.buildSpec.url)
+    this.chmodDone = false
   }
 
   async getPath(log: LogEntry) {
     await this.download(log)
```
Comment (eysi09) [inline garden-service/src/util/ext-tools.ts:107]: See my comment above about downloading here. I personally would expect `getPath` to just return null and not have side effects if the path is not found. But again, not part of the PR per se, just flagging in case you agree.

File: garden-service/src/util/ext-tools.ts [hunk 2]
```diff
@@ -7,93 +7,193 @@
...
-    return join(this.versionPath, ...this.targetSubpath)
+    const path = join(this.versionPath, ...this.targetSubpath.split(posix.sep))
+
+    if (this.spec.type === "binary") {
+      // Make sure the target path is executable
+      if (!this.chmodDone) {
+        await chmod(path, 0o755)
```
Comment (eysi09) [inline garden-service/src/util/ext-tools.ts:113]: This won't work on Windows. Should we just expect the user to set the permissions manually in that case?

Comment (edvald) [reply garden-service/src/util/ext-tools.ts:113]: It won't? It's been working so far.

File: garden-service/src/util/ext-tools.ts (hunk 1 above)
Comment (edvald) [reply garden-service/src/util/ext-tools.ts:107]: I think this is fine. Returning null here just pushes the user to check for the null and then download.

File: garden-service/src/util/ext-tools.ts [hunk 3]
```diff
@@ -7,93 +7,193 @@
...
-  // Optionally specify sha256 checksum for validation.
-  sha256?: string
-  // If the URL contains an archive, provide extraction instructions.
-  extract?: LibraryExtractSpec
+export interface ExecParams {
+  args?: string[]
+  cwd?: string
+  env?: { [key: string]: string }
+  log: LogEntry
+  timeout?: number
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/util/ext-tools.ts:38]: This is unchanged from previous code, but yeah I wouldn't mind making the unit explicit everywhere.

File: garden-service/src/commands/tools.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+      if (plugins.length === 0) {
+        throw new ParameterError(`Could not find plugin ${pluginName}.`, { availablePlugins })
+      }
+    } else {
+      // Place configured providers at the top for preference, if applicable
+      const projectRoot = await findProjectConfig(garden.projectRoot)
+
+      if (projectRoot) {
+        if (garden instanceof DummyGarden) {
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/commands/tools.ts:131]: Not when unit testing, needed to account for that. A comment wouldn't hurt to explain that though.

File: garden-service/src/commands/tools.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  arguments = toolsArgs
+  options = toolsOpts
+
+  getLoggerType(): LoggerType {
+    return "basic"
+  }
+
+  async prepare({ log }) {
+    // Override the logger output, to output to stderr instead of stdout, to avoid contaminating command output
+    log.root.writers.find((w) => w.type === "basic")!.output = process.stderr
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/commands/tools.ts:89]: It's common to parse stdout from commands. Sure, the called command may write to stderr, but that's almost never meant to be machine readable. This helps maintain that separation. Say if a command is expected to return a JSON object, writing our log output to stdout would make it unusable.

File: garden-service/src/commands/tools.ts [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+type Args = typeof toolsArgs
+type Opts = typeof toolsOpts
+
+export class ToolsCommand extends Command<Args, Opts> {
+  name = "tools"
+  help = "Access tools included by providers."
+  cliOnly = true
+
+  // FIXME: We need this while we're still resolving providers in the AnalyticsHandler
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/commands/tools.ts:45]: We do now, not when this was written originally. You fixed that quite recently. Comment is out of date, but the flag remains necessary, so I'll remove the comment.

File: docs/reference/commands.md [hunk 7]
```diff
@@ -904,41 +982,36 @@ Examples:
 
     garden update-remote all 
 
-### garden validate
+### garden util fetch-tools
```
[the remark this answers is not in the document]
Comment (edvald) [reply docs/reference/commands.md:985]: I considered that, but I felt that would be conflating use-cases and be more confusing, both in terms of docs and code.

File: garden-service/src/util/ext-tools.ts (hunk 2 above)
Comment (eysi09) [reply garden-service/src/util/ext-tools.ts:113]: Hmm ok. I Googled this when reviewing and found: https://stackoverflow.com/questions/8756639/how-do-i-use-chmod-with-node-js

...and went on the assumption that this was just a no-op on Windows.

File: garden-service/src/analytics/analytics.ts [hunk 8]
```diff
@@ -247,7 +243,7 @@ export class AnalyticsHandler {
    * Used internally to check if a users has opted-in or not.
    */
   private analyticsEnabled(): boolean {
-    if (process.env.GARDEN_DISABLE_ANALYTICS === "true") {
+    if (!this.sessionId || process.env.GARDEN_DISABLE_ANALYTICS === "true") {
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/analytics/analytics.ts:246]: Is there a valid scenario where the session ID is null but we want events to be sent?

File: garden-service/src/analytics/analytics.ts [hunk 9]
```diff
@@ -126,15 +125,12 @@ export class AnalyticsHandler {
   private ciName = ci.name
   private systemConfig: SystemInfo
   private isCI = ci.isCI
-  private sessionId: string
+  private sessionId: string | null
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/analytics/analytics.ts:128]: I kept getting errors in unit tests. As far as I can tell this case (null sessionId) only comes up there. I'm open to a separate fix for that, but I couldn't quite grok why this was how it was so I'd ask you or @thsig to investigate.

File: garden-service/src/util/ext-tools.ts (hunk 2 above)
Comment (edvald) [reply garden-service/src/util/ext-tools.ts:113]: If it is, then that's fine, and just not needed. It's been working fine for many releases, no change in behavior here.

File: docs/reference/commands.md [hunk 10]
```diff
@@ -793,6 +821,56 @@ Examples:
...
+directory (depending on the tool type). If the tool is a non-executable library, this
+flag is implicit.
+
+When multiple plugins provide a tool with the same name, you can choose a specific
+plugin/version by specifying <plugin name>.<tool name>, instead of just <tool name>.
+This is generally advisable when using this command in scripts, to avoid accidental
+conflicts.
+
+When there are name conflicts and a plugin name is not specified, the preference is
+for defined by configured providers in the current project (if applicable), and then
```
[the remark this answers is not in the document]
Comment (edvald) [reply docs/reference/commands.md:841]: Actually it's supposed to be "the preference is for tools defined by configured providers". I'll fix and likely rephrase for better clarity.

**Comment by edvald** [comment]
> Are we verifying anywhere, either explicitly or implicitly, that the tool remote URLs are correct across all platforms?

That would be good. We haven't done that so far, but it would be feasible to create an integ test suite that does that, or something in CI.

> Is there any concern that the tools will conflict with previous installations? For example, if you have two Docker installations or install a tool via Chocolate and again via Garden.

Not really, those won't conflict. But we will indeed fetch Docker even if it's already installed.

> You mentioned the images where too bloated, how do we plan on addressing that?

I added a couple of to-do items in our 0.12 checklist, for dropping old tools that we don't aim to support anymore. That's a big chunk. Then there are smaller improvements we could make, e.g. make it possible to drop files that we don't use in archives we extract.

File: garden-service/src/analytics/analytics.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (eysi09) [reply garden-service/src/analytics/analytics.ts:128]: Yeah the `sessionId` appears to be created in the `CLI` class and passed to `BufferedEventStream` and the `Garden` class. It doesn't look its really used by the `Garden` class though, so perhaps we could just pass it directly to the `AnalyticsHandler` class.

In any case, I don't think we should change the behaviour of the `AnalyticsHandler` class to accommodate the unit tests.

File: garden-service/src/analytics/analytics.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (eysi09) [reply garden-service/src/analytics/analytics.ts:246]: IMO the issue is that the `sessionId` shouldn't be null to begin with. That's why before this commit, the `AnalyticsHandler` would just throw if it was missing.

File: garden-service/src/analytics/analytics.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/analytics/analytics.ts:128]: If the `AnalyticsHandler` is never supposed to receive a null sessionId, it should imo be caught at compile time. I'm unclear on why there's a runtime guard there then.

[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/analytics/analytics.ts:128]: To be clear, I wasn't trying to get _my_ tests to pass, other unrelated tests kept failing. I can revert this, but then we'd need a separate PR to fix the failing tests.

**Comment by eysi09** [comment]
Thanks for the answers, make sense. Do you want to look into the tests now, or just the next time we actually tamper with the tool URLs?

**Comment by edvald** [comment]
We already test most of them implicitly in container builds, integ/e2e tests and Windows tests, so I'm not sure I'd want to get into more explicit tool tests. Until a new reason comes up, that is.
